From 15b1a3dfc84be620cc9ca5f34a32bc0012ef26fa Mon Sep 17 00:00:00 2001
From: Paulo Sergio Travaglia <pstglia@gmail.com>
Date: Tue, 21 Apr 2015 22:03:00 -0300
Subject: [PATCH] Enable radeonsi on Android-x86 (lollipop)

These changes enables radeonsi compiling on Android-x86.

The main modifications are:

- Include paths for LLVM header files and shared/static libraries
- Set C++ flag "c++11" to avoid compiling errors on LLVM header files
- Set defines for LLVM (-D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -DHAVE_LLVM=0x0305 -DLLVM_VERSION_PATCH=2)
  (Note: llvm on Android seems a mix of 3.4 and 3.5. If set to 3.4, complains about missing include files...)
- Add GALLIVM source files
- Changes path of libelf library for lollipop
---
 Android.common.mk                         |  9 +++++++--
 Android.mk                                |  1 +
 src/egl/drivers/dri2/Android.mk           |  3 ++-
 src/egl/main/Android.mk                   |  8 +++++++-
 src/gallium/auxiliary/Android.mk          | 13 +++++++++++--
 src/gallium/drivers/radeon/Android.mk     |  7 ++++++-
 src/gallium/state_trackers/egl/Android.mk |  3 ++-
 src/util/Android.mk                       | 30 ++++++++++++++++--------------
 8 files changed, 52 insertions(+), 22 deletions(-)

diff --git a/Android.common.mk b/Android.common.mk
index 26c91f5..5aa7eb1 100644
--- a/Android.common.mk
+++ b/Android.common.mk
@@ -39,7 +39,9 @@ LOCAL_CFLAGS += \
 	-DHAVE___BUILTIN_UNREACHABLE \
 	-DPACKAGE_VERSION=\"$(MESA_VERSION)\" \
 	-DPACKAGE_BUGREPORT=\"https://bugs.freedesktop.org/enter_bug.cgi?product=Mesa\" \
-	-DANDROID_VERSION=0x0$(MESA_ANDROID_MAJOR_VERSION)0$(MESA_ANDROID_MINOR_VERSION)
+	-DANDROID_VERSION=0x0$(MESA_ANDROID_MAJOR_VERSION)0$(MESA_ANDROID_MINOR_VERSION) \
+	-DHAVE_LLVM=0x0305 -DLLVM_VERSION_PATCH=2 \
+	-D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
 
 LOCAL_CFLAGS += \
 	-DHAVE_PTHREAD=1 \
@@ -57,7 +59,10 @@ endif
 
 LOCAL_CPPFLAGS += \
 	-Wno-error=non-virtual-dtor \
-	-Wno-non-virtual-dtor
+	-Wno-non-virtual-dtor \
+	-DHAVE_LLVM=0x0305 -DLLVM_VERSION_PATCH=2 \
+	-D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
+
 
 # uncomment to keep the debug symbols
 #LOCAL_STRIP_MODULE := false
diff --git a/Android.mk b/Android.mk
index b81c060..4b8f507 100644
--- a/Android.mk
+++ b/Android.mk
@@ -40,6 +40,7 @@ MESA_PYTHON2 := python
 
 DRM_TOP := external/drm
 DRM_GRALLOC_TOP := hardware/drm_gralloc
+LLVM_TOP := external/llvm
 
 classic_drivers := i915 i965
 gallium_drivers := swrast freedreno i915g ilo nouveau r300g r600g radeonsi vmwgfx
diff --git a/src/egl/drivers/dri2/Android.mk b/src/egl/drivers/dri2/Android.mk
index 984dcf0..738c219 100644
--- a/src/egl/drivers/dri2/Android.mk
+++ b/src/egl/drivers/dri2/Android.mk
@@ -41,7 +41,8 @@ LOCAL_C_INCLUDES := \
 	$(MESA_TOP)/src/egl/main \
 	$(MESA_TOP)/src/loader \
 	$(DRM_TOP)/include/drm \
-	$(DRM_GRALLOC_TOP)
+	$(DRM_GRALLOC_TOP) \
+	system/core/libsync/include
 
 LOCAL_STATIC_LIBRARIES := \
 	libmesa_loader
diff --git a/src/egl/main/Android.mk b/src/egl/main/Android.mk
index e3d6f8b..d283dd8 100644
--- a/src/egl/main/Android.mk
+++ b/src/egl/main/Android.mk
@@ -43,7 +43,11 @@ LOCAL_CFLAGS := \
 	-D_EGL_DRIVER_SEARCH_DIR=\"/system/lib/egl\" \
 	-D_EGL_OS_UNIX=1
 
-LOCAL_STATIC_LIBRARIES :=
+LOCAL_STATIC_LIBRARIES := \
+	libLLVMR600CodeGen \
+	libLLVMR600Desc \
+	libLLVMR600Info \
+	libLLVMR600AsmPrinter \
 
 LOCAL_SHARED_LIBRARIES := \
 	libglapi \
@@ -52,6 +56,7 @@ LOCAL_SHARED_LIBRARIES := \
 	liblog \
 	libcutils \
 	libgralloc_drm \
+	libLLVM \
 
 ifeq ($(shell echo "$(MESA_ANDROID_VERSION) >= 4.2" | bc),1)
 LOCAL_SHARED_LIBRARIES += libsync
@@ -148,6 +153,7 @@ LOCAL_STATIC_LIBRARIES := \
 	libmesa_glsl \
 	libmesa_glsl_utils \
 	libmesa_gallium \
+	libelf \
 	$(LOCAL_STATIC_LIBRARIES)
 
 endif # MESA_BUILD_GALLIUM
diff --git a/src/gallium/auxiliary/Android.mk b/src/gallium/auxiliary/Android.mk
index 8046943..f876d0b 100644
--- a/src/gallium/auxiliary/Android.mk
+++ b/src/gallium/auxiliary/Android.mk
@@ -28,11 +28,20 @@ include $(LOCAL_PATH)/Makefile.sources
 
 include $(CLEAR_VARS)
 
-LOCAL_SRC_FILES := $(C_SOURCES)
+LOCAL_CPPFLAGS += \
+  -std=c++11
+
+LOCAL_SRC_FILES := \
+	$(C_SOURCES) \
+	$(GALLIVM_SOURCES) \
+	$(GALLIVM_CPP_SOURCES)
 
 LOCAL_C_INCLUDES := \
 	$(GALLIUM_TOP)/auxiliary/util \
-	$(MESA_TOP)/src
+	$(MESA_TOP)/src \
+	$(LLVM_TOP)/include \
+	$(LLVM_TOP)/device/include \
+	external/libcxx/include
 
 LOCAL_MODULE := libmesa_gallium
 
diff --git a/src/gallium/drivers/radeon/Android.mk b/src/gallium/drivers/radeon/Android.mk
index 7220d17..b41f58a 100644
--- a/src/gallium/drivers/radeon/Android.mk
+++ b/src/gallium/drivers/radeon/Android.mk
@@ -28,10 +28,15 @@ include $(LOCAL_PATH)/Makefile.sources
 
 include $(CLEAR_VARS)
 
-LOCAL_SRC_FILES := $(C_SOURCES)
+LOCAL_SRC_FILES := \
+	$(C_SOURCES) \
+	$(LLVM_C_FILES)
 
 LOCAL_C_INCLUDES := \
 	$(DRM_TOP)/radeon \
+	$(LLVM_TOP)/include \
+	$(LLVM_TOP)/device/include \
+	external/elfutils/0.153/libelf \
 
 LOCAL_MODULE := libmesa_pipe_radeon
 
diff --git a/src/gallium/state_trackers/egl/Android.mk b/src/gallium/state_trackers/egl/Android.mk
index d198762..3c013ec 100644
--- a/src/gallium/state_trackers/egl/Android.mk
+++ b/src/gallium/state_trackers/egl/Android.mk
@@ -40,7 +40,8 @@ LOCAL_CFLAGS := \
 LOCAL_C_INCLUDES := \
 	$(GALLIUM_TOP)/state_trackers/egl \
 	$(GALLIUM_TOP)/winsys \
-	$(MESA_TOP)/src/egl/main
+	$(MESA_TOP)/src/egl/main \
+	system/core/libsync/include
 
 # swrast only
 ifeq ($(MESA_GPU_DRIVERS),swrast)
diff --git a/src/util/Android.mk b/src/util/Android.mk
index 2eb0ef7..2ad27d7 100644
--- a/src/util/Android.mk
+++ b/src/util/Android.mk
@@ -31,7 +31,8 @@ include $(LOCAL_PATH)/Makefile.sources
 include $(CLEAR_VARS)
 
 LOCAL_SRC_FILES := \
-	$(MESA_UTIL_FILES)
+	$(MESA_UTIL_FILES) \
+	$(MESA_UTIL_GENERATED_FILES)
 
 LOCAL_C_INCLUDES := \
 	$(MESA_TOP)/src/mesa \
@@ -44,13 +45,13 @@ LOCAL_MODULE := libmesa_util
 
 LOCAL_MODULE_CLASS := STATIC_LIBRARIES
 
-intermediates := $(call local-intermediates-dir)
-LOCAL_GENERATED_SOURCES := $(addprefix $(intermediates)/,$(MESA_UTIL_GENERATED_FILES))
+#intermediates := $(call local-intermediates-dir)
+#LOCAL_GENERATED_SOURCES := $(addprefix $(intermediates)/,$(MESA_UTIL_GENERATED_FILES))
 
-$(LOCAL_GENERATED_SOURCES): PRIVATE_PYTHON := $(MESA_PYTHON2)
-$(LOCAL_GENERATED_SOURCES): PRIVATE_CUSTOM_TOOL = $(PRIVATE_PYTHON) $^ > $@
-$(LOCAL_GENERATED_SOURCES): $(intermediates)/%.c: $(LOCAL_PATH)/%.py
-	$(transform-generated-source)
+#$(LOCAL_GENERATED_SOURCES): PRIVATE_PYTHON := $(MESA_PYTHON2)
+#$(LOCAL_GENERATED_SOURCES): PRIVATE_CUSTOM_TOOL = $(PRIVATE_PYTHON) $^ > $@
+#$(LOCAL_GENERATED_SOURCES): $(intermediates)/%.c: $(LOCAL_PATH)/%.py
+#	$(transform-generated-source)
 
 include $(MESA_COMMON_MK)
 include $(BUILD_STATIC_LIBRARY)
@@ -65,7 +66,8 @@ LOCAL_IS_HOST_MODULE := true
 LOCAL_CFLAGS := -D_POSIX_C_SOURCE=199309L
 
 LOCAL_SRC_FILES := \
-	$(MESA_UTIL_FILES)
+	$(MESA_UTIL_FILES) \
+	$(MESA_UTIL_GENERATED_FILES)
 
 LOCAL_C_INCLUDES := \
 	$(MESA_TOP)/src/mesa \
@@ -78,13 +80,13 @@ LOCAL_MODULE := libmesa_util
 
 LOCAL_MODULE_CLASS := STATIC_LIBRARIES
 
-intermediates := $(call local-intermediates-dir)
-LOCAL_GENERATED_SOURCES := $(addprefix $(intermediates)/,$(MESA_UTIL_GENERATED_FILES))
+#intermediates := $(call local-intermediates-dir)
+#LOCAL_GENERATED_SOURCES := $(addprefix $(intermediates)/,$(MESA_UTIL_GENERATED_FILES))
 
-$(LOCAL_GENERATED_SOURCES): PRIVATE_PYTHON := $(MESA_PYTHON2)
-$(LOCAL_GENERATED_SOURCES): PRIVATE_CUSTOM_TOOL = $(PRIVATE_PYTHON) $^ > $@
-$(LOCAL_GENERATED_SOURCES): $(intermediates)/%.c: $(LOCAL_PATH)/%.py
-	$(transform-generated-source)
+#$(LOCAL_GENERATED_SOURCES): PRIVATE_PYTHON := $(MESA_PYTHON2)
+#$(LOCAL_GENERATED_SOURCES): PRIVATE_CUSTOM_TOOL = $(PRIVATE_PYTHON) $^ > $@
+#$(LOCAL_GENERATED_SOURCES): $(intermediates)/%.c: $(LOCAL_PATH)/%.py
+#	$(transform-generated-source)
 
 include $(MESA_COMMON_MK)
 include $(BUILD_HOST_STATIC_LIBRARY)
-- 
1.7.12.1

